src/NuGet.Services.AzureSearch/Db2AzureSearch/Db2AzureSearchCommand.cs

```diff
@@ -0,0 +1,228 @@
+            try
+            {
+                while ((allWork.TryTake(out work) || !completedToken.IsCancellationRequested)
+                    && !cancellationToken.IsCancellationRequested)
+                {
+                    // If there's no work to do, wait a bit before checking again.
+                    if (work == null)
+                    {
+                        await Task.Delay(TimeSpan.FromMilliseconds(100));
+                        continue;
+                    }
+
+                    var indexActions = _indexActionBuilder.AddNewPackageRegistration(work);
+
+                    foreach (var action in indexActions.Search)
+                    {
+                        searchActions.Enqueue(action);
+                    }
+
+                    foreach (var action in indexActions.Hijack)
+                    {
+                        hijackActions.Enqueue(action);
+                    }
+
+                    var searchBatchesTask = PushFullBatchesAsync(_searchIndexClient, searchActions);
+                    var hijackBatchesTask = PushFullBatchesAsync(_hijackIndexClient, hijackActions);
+                    await Task.WhenAll(searchBatchesTask, hijackBatchesTask);
+                }
+
+                var searchFinishTask = IndexAsync(_searchIndexClient, searchActions);
+                var hijackFinishTask = IndexAsync(_hijackIndexClient, hijackActions);
+                await Task.WhenAll(searchFinishTask, hijackFinishTask);
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(
+                    0,
+                    ex,
+                    "An exception was thrown while processing package ID {PackageId}.",
+                    work?.PackageId ?? "(last batch...)");
+                throw;
+            }
+        }
+
+        private async Task PushFullBatchesAsync(
+            ISearchIndexClientWrapper indexClient,
+            Queue<IndexAction<KeyedDocument>> actions)
+        {
+            while (actions.Count > _options.Value.AzureSearchBatchSize)
+            {
+                var batch = new List<IndexAction<KeyedDocument>>();
+                while (batch.Count < _options.Value.AzureSearchBatchSize)
+                {
+                    batch.Add(actions.Dequeue());
+                }
+
+                await IndexAsync(indexClient, batch);
+            }
+        }
+
+        private async Task IndexAsync(
+            ISearchIndexClientWrapper indexClient,
+            IReadOnlyCollection<IndexAction<KeyedDocument>> batch)
+        {
+            if (batch.Count == 0)
+            {
+                return;
+            }
+
+            await Task.Yield();
+
+            _logger.LogInformation(
+                "Pushing batch of {BatchSize} to index {IndexName}.",
+                batch.Count,
+                indexClient.IndexName);
+            var batchResults = await indexClient.Documents.IndexAsync(new IndexBatch<KeyedDocument>(batch));
```

Comment:
Is there a need to pass a `new IndexBatch`? Does `Documents.IndexAsync` not treat the passed batch as readonly?

Reply:
`IndexAsync` requires an `IndexBatch` -- not `List` or `IEnumerable`.